lib/tls.js

```diff
@@ -321,6 +321,14 @@ function onCryptoStreamEnd() {
   }
 
   if (this.onend) this.onend();
+
+  if (this._opposite._finished &&
+      !this._opposite._ended &&
+      !this._opposite._doneFlag) {
+    this._opposite.once('end', function(xs) {
+      this._done();
```

Comment:
Could you please paste a `console.trace()` at this place when the bug manifests itself? I suppose that it comes from the shutdown handling line.